source/reflect_impl_load_metafunction.h

```diff
@@ -0,0 +1,165 @@
+    {
+#ifdef _WIN32
+        auto symbol = GetProcAddress(static_cast<HMODULE>(handle_), name.c_str());
+#else
+        auto symbol = dlsym(handle_, name.c_str());
+        if(!symbol)
+        {
+            // Some platforms export with additional underscore, so try that.
+            auto const us_name = "_" + name;
+            symbol = dlsym(handle_, us_name.c_str());
+        }
+#endif // _WIN32
+        return function_cast_<T*>(symbol);
+    }
+private:
+    void* handle_{nullptr};
+
+    template<typename T>
+    static auto function_cast_(auto ptr) noexcept -> T
+    {
+        using generic_function_ptr = void (*)();
+        return reinterpret_cast<T>(reinterpret_cast<generic_function_ptr>(ptr));
+    }
+
+    static auto get_last_error_() noexcept -> std::string
+    {
+#ifdef _WIN32
+        DWORD errorMessageID = GetLastError();
+        if(errorMessageID == 0) {
+            return {}; // No error message has been recorded
+        }
+        LPSTR messageBuffer = nullptr;
+        auto size = FormatMessageA(
+            FORMAT_MESSAGE_ALLOCATE_BUFFER|FORMAT_MESSAGE_FROM_SYSTEM|FORMAT_MESSAGE_IGNORE_INSERTS,
+            nullptr,
+            errorMessageID,
+            MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US),
+            (LPSTR)&messageBuffer,
+            0,
+            nullptr
+        );
+        std::string message(messageBuffer, static_cast<size_t>(size));
+        LocalFree(messageBuffer);
+        return message;
+#else
+        return std::string{dlerror()};
+#endif // _WIN32
+    }
+
+};
+
+
+struct load_metafunction_ret {
+    std::function<void(type_declaration&)> metafunction;
+    std::string error;
+};
+
+
+//  Load metafunction by opening DLL with OS APIs
+//
+//  The environment variable 'CPPFRONT_METAFUNCTION_LIBRARIES'
+//  is read and interpreted as ':'-separated library paths
+//  where a metafunction symbol is looked up at
+auto load_metafunction(std::string const& name)
+    -> load_metafunction_ret
+{
+    // FIXME: On Windows, using this approach with the system apis not set to utf8, will
+    // break if a metafunction library contains unicode codepoints in its name, a proper
+    // way to handle this would be to use _wgetenv and use wchar_t strings for the dll opening
+    // function
+    auto cpp1_libraries_cstr = std::getenv("CPPFRONT_METAFUNCTION_LIBRARIES");
+    if (!cpp1_libraries_cstr) {
+        return {};
+    }
+
+    auto cpp1_libraries = std::string_view{cpp1_libraries_cstr};
+    if (cpp1_libraries.empty()) {
+        return {};
+    }
+
+    auto cpp1_name = "cpp2_metafunction_" + name;
+    while (!cpp1_libraries.empty())
+    {
+        auto colon = cpp1_libraries.find(':');
+        auto lib_path = cpp1_libraries.substr(0, colon);
+        cpp1_libraries.remove_prefix(lib_path.size() + unsigned(colon != lib_path.npos));
+
+        auto lib = std::make_shared<dll>(std::string(lib_path));
+        if(!lib->is_open()) {
+            continue;
```

Comment:
I have a question with regards to defensive programming.

Since commit 1bc650691c1289bb9c99e6068ae91203dfcc4a3e, the successful opening of the library is contract checked.
`is_open` is "no longer necessary", but I forgot to excise it, and to maybe defensively assert `handler_` in `get_alias`.
But actually, execution reaches this point if contract evaluation is turned off (commit 2893e64cae356f01721deede8d488dc7e47a3b4c).
Then, in Linux, we would be at the mercy of [`dlsym`](https://www.man7.org/linux/man-pages/man3/dlsym.3.html).
`dlsym` is documented as taking a "handle" to the DLL returned by `dlopen`, which `nullptr` certainly is not.

What's the right approach here?
I suppose it's a matter of culture.
The code of cppfront at large seems to be defensively checked with Cpp1 `assert` (effective in Debug builds).
And by its nature, its defensively guarded with `if`s.

I personally prefer to avoid defensive programming and assume that my contracts are always evaluated.
That also includes not turning off the checked-by default Cpp2 code
(with `-no-comparison-checks -no-null-checks -no-subscript-checks -fno-exceptions -fno-rtti`).
That results in better flowing code, pretty much as when using exceptions.
For my use of contracts, maybe I should be using a contract group that can't turn off evaluation.
Unfortunately (?), the built-in safety of Cpp2 uses contract groups that can be turned off.

Reply:
If I were to use this on my own codebase I would not even let the constructor complete, I'd raise an exception in case the handle is not valid, with an explanatory message, which would elide completely the need for `is_open` (RAII FTW). I didn't see exception handling in cppfront which is why I decided to go with this approach instead, but if you added a check that is going to stop execution as soon as there is an error, then I guess that has the same purpose as the `is_open` handling?

Reply:
Thank you, that got me thinking.
In this case, <https://github.com/hsutter/cppfront/pull/831#issuecomment-1858928061> is the answer.
I should use the `meta::type_declaration` as the contract group.
I considered it briefly, but I wanted to keep the surface area of the header minimal.
I will take this as cue to move everything into `reflect_impl.h2`.